### PR TITLE
SCRAM-SHA-256 support

### LINKED ISSUE #255: Implement SCRAM-SHA-256 authentication scheme
Our corporate database servers, it turns out, are requiring this.  I'll take a crack at doing it in the next couple weeks if no one else has plans to.

### PR DESCRIPTION
**Pull request by mpilquist**
resolves #255

### CODE REVIEW

File: modules/core/src/main/scala/net/message/Scram.scala [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    s"n=,r=${nonce}".bytesUtf8
+
+  case class ServerFirst(nonce: String, salt: ByteVector, iterations: Int)
+  object ServerFirst {
+    private val Pattern = """r=([\x21-\x2B\x2D-\x7E]+),s=([A-Za-z0-9+/]+={0,2}),i=(\d+)""".r
+
+    def decode(bytes: ByteVector): Option[ServerFirst] =
+      utf8.decodeValue(bytes.bits).toOption.flatMap {
+        case Pattern(r, s, i) =>
+          Some(ServerFirst(r, ByteVector.fromBase64(s).get, i.toInt))
```
Comment (sbuzzard) [inline modules/core/src/main/scala/net/message/Scram.scala:43]: isn’t there a fromValidBase64?

File: modules/core/src/main/scala/net/message/Scram.scala [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  case class Verifier(value: ByteVector)
+
+  case class ServerFinal(verifier: Verifier)
+  object ServerFinal {
+    private val Pattern = """v=([A-Za-z0-9+/]+={0,2})""".r
+    def decode(bytes: ByteVector): Option[ServerFinal] =
+      utf8.decodeValue(bytes.bits).toOption.flatMap {
+        case Pattern(v) =>
```
Comment (sbuzzard) [inline modules/core/src/main/scala/net/message/Scram.scala:66]: same. i know, the tiniest of nits but seeing get even when its impossible to fail ... -)

File: modules/core/src/main/scala/net/message/Scram.scala [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    com.ongres.saslprep.SaslPrep.saslPrep(value, false)
+
+  def clientFirstBare: ByteVector = {
+    val nonce = {
+      val arr = new Array[Byte](32)
+      java.security.SecureRandom.getInstanceStrong().nextBytes(arr)
+      ByteVector.view(arr).toBase64
+    }
+    clientFirstBareWithNonce(nonce)
+  }
```
Comment (sbuzzard) [inline modules/core/src/main/scala/net/message/Scram.scala:31]: i suppose for this kind of thing purity isn’t the point but this i assume isn’t pure with the secure random call - would we want to wrap in an effect?

**Review by sbuzzard — APPROVED** [review]

File: modules/core/src/main/scala/net/message/Scram.scala (hunk 3 above)
Comment (mpilquist) [reply modules/core/src/main/scala/net/message/Scram.scala:31]: I don't know - I originally had this in a `F[_]: Sync` but that percolates up the call stack. Adding a `Random` effect seemed overkill. So I simplified to this version.

File: build.sbt [hunk 4]
```diff
@@ -81,6 +81,7 @@ lazy val core = project
       "org.scodec"    %% "scodec-cats"  % "1.0.0",
       "com.beachape"  %% "enumeratum"   % "1.6.1",
       "org.tpolecat"  %% "natchez-core" % "0.0.12",
+      "com.ongres.stringprep" % "saslprep" % "1.1"
```
Comment (mpilquist) [inline build.sbt:84]: This is BSD 2-clause licensed and seems safe / unlikely to change or cause dependency issues for folks downstream. The main algorithm is here, which is 100+ lines of unicode manipulation: https://gitlab.com/ongresinc/stringprep/-/blob/development/saslprep/src/main/java/com/ongres/saslprep/SaslPrep.java

Comment (tpolecat) [reply build.sbt:84]: :+1: seems fine

File: modules/core/src/main/scala/net/message/Scram.scala [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    val mac = javax.crypto.Mac.getInstance("HmacSHA256")
+    mac.init(new javax.crypto.spec.SecretKeySpec(key.toArray, "HmacSHA256"))
+    ByteVector.view(mac.doFinal(str.toArray))
+  }
+
+  private def H(input: ByteVector): ByteVector = input.digest("SHA-256")
+
+  private def Hi(str: String, salt: ByteVector, iterations: Int): ByteVector = {
+    val spec = new javax.crypto.spec.PBEKeySpec(str.toCharArray, salt.toArray, iterations, 8 * 32)
+    val salted = javax.crypto.SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256").generateSecret(spec).getEncoded
```
Comment (tpolecat) [inline modules/core/src/main/scala/net/message/Scram.scala:85]: Shit almighty man, JCA gives me the heebiejeebies.

**Review by tpolecat — APPROVED** [review]
Amazing contribution, thank you!

**Comment by tpolecat** [comment]
tagged  v0.0.20

**Pull request merged**